controllers/release/release_adapter_test.go

```diff
@@ -0,0 +1,414 @@
+			return (!result.CancelRequest && err == nil)
+		}).Should(BeTrue())
+		pipelineRun, _ := cacheAdapter.getReleasePipelineRun()
+		Expect(pipelineRun).NotTo(BeNil())
+
+		result, err := cacheAdapter.EnsureReleasePipelineStatusIsTracked()
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+
+		// should return results.ContinueProcessing() in case the release is done
+		release.MarkSucceeded()
+		result, err = cacheAdapter.EnsureReleasePipelineStatusIsTracked()
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+
+		// release is running but has no pipelineRun matching the release
+		release.MarkRunning()
+		pipelineRun.ObjectMeta.Labels = nil
+		cacheClient.Update(ctx, pipelineRun)
+		release.Status.Conditions = []metav1.Condition{}
+		result, err = cacheAdapter.EnsureReleasePipelineStatusIsTracked()
+
+		// err should not be nil here but cacheAdapter.getReleasePipelineRun does not advise
+		// there is no pipelineRun for the release
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+
+		// local cleanup
+		cacheClient.Delete(ctx, pipelineRun)
+	})
+
+	It("can make sure Finalizers are called for a Release object", func() {
+		operations := []ReconcileOperation{
+			cacheAdapter.EnsureFinalizersAreCalled,
+			cacheAdapter.EnsureReleasePipelineRunExists,
+			cacheAdapter.EnsureReleasePipelineStatusIsTracked,
+		}
+
+		for _, operation := range operations {
+			result, err := operation()
+			Expect(!result.CancelRequest && err == nil).To(BeTrue())
+		}
+		release.SetDeletionTimestamp(&metav1.Time{Time: time.Now()})
+
+		// should Requeue if there are no Finalizers
+		result, err := cacheAdapter.EnsureFinalizersAreCalled()
+		Expect(result.RequeueRequest == true && err == nil).To(BeTrue())
+
+		// adding finalizer - should err as DeletionTimestamp was set, hence
+		// err is ignored.
+		result, _ = cacheAdapter.EnsureFinalizerIsAdded()
+		Expect(!result.CancelRequest).To(BeTrue())
+
+		cacheClient.Delete(ctx, release)
+
+		result, err = cacheAdapter.EnsureFinalizersAreCalled()
+		// err should occur as it is not possible to update the immutable
+		// DeletionTimestamp of the object, so it will throw an error, but
+		// this is ok for the
+		Expect(!result.CancelRequest && err != nil).To(BeTrue())
+
+		// if all worked as expected, the pipelineRun should be gone.
+		pipelineRun, _ := cacheAdapter.getReleasePipelineRun()
+		Expect(cacheClient.Delete(ctx, pipelineRun)).ShouldNot(Succeed())
+	})
+
+	It("can make sure Finalizers are called and fails for a Release object when the pipelineRun was already finalized", func() {
+		operations := []ReconcileOperation{
+			cacheAdapter.EnsureFinalizersAreCalled,
+			cacheAdapter.EnsureReleasePipelineRunExists,
+			cacheAdapter.EnsureFinalizerIsAdded,
+			cacheAdapter.EnsureReleasePipelineStatusIsTracked,
+		}
+		for _, operation := range operations {
+			result, err := operation()
+			Expect(!result.CancelRequest && err == nil).To(BeTrue())
+		}
```

Comment:
I´d say that you only need an eventually block to check that the finalizer is there and then do your stuff.

Reply:
Ok. Now I got what you meant  in the other comment regarding this kind of block.

Reply:
Fixed.